src/Core/Logging/LoggerBase.cs

```diff
@@ -0,0 +1,94 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Globalization;
+using System.Text;
+
+namespace SonarLint.VisualStudio.Core.Logging;
+
+internal class LoggerBase(
+    ILoggerContextManager contextManager,
+    ILoggerWriter writer,
+    ILoggerSettingsProvider settingsProvider) : ILogger
+{
+    public ILogger ForContext(params string[] context) =>
```

Comment:
Creating a new instance for each context feels like a very bad design decision as this will increase the memory usage and the performance overhead.

Reply:
This is not a correct conclusion. The instances are lightweight (partially thanks to immutable lists and lazy string property initialization, but in general they don't hold a lot of data apart from a few object references and the aforementioned property list and formatted string). This will not increase the overhead in any meaningful way.

Reply:
On top of that, most of these instances will be created in user-class constructors and therefore will be long-living.

Reply:
It is a correct conclusion and it will bring more pressure to the garbage collector, which will impact the performance.

> On top of that, most of these instances will be created in user-class constructors and therefore will be long-living.
This is correct for the LoggerBase instances. That's why I see no reason why it shouldn't be a Mef component with a non-shared creation policy (which will basically lead to the same thing but without the need of factories).
This is not true, however for the LoggerContextManager and I think we can find better design than just creating instances for each context.

Reply:
Again, that's not true. It will, of course impact the performance, as pretty much every single added allocation or instruction does, but the question is will it impact it in any meaningful way? The answer is not. It has less of an impact on memory than the existing string allocations inside the logger.

Reply:
Uncustomized logger context manager has references to 2 empty immutable lists and 2 non-initialized lazies...